README.md
[![Clojars Project](https://img.shields.io/clojars/v/clj-bson-rpc.svg)](https://clojars.org/clj-bson-rpc)
[![Build Status](https://travis-ci.org/seprich/clj-bson-rpc.svg?branch=master)](https://travis-ci.org/seprich/clj-bson-rpc)
[![Dependencies Status](https://jarkeeper.com/seprich/clj-bson-rpc/status.svg)](https://jarkeeper.com/seprich/clj-bson-rpc)


# clj-bson-rpc


A Clojure library for JSON-RPC 2.0 and BSON-RPC on TCP (+ TLS).

| [API doc](http://seprich.github.io/clj-bson-rpc/codox/clj-bson-rpc.core.html)
| [Marginalia](http://seprich.github.io/clj-bson-rpc/marginalia.html)
| [Change Log](https://github.com/seprich/clj-bson-rpc/blob/master/CHANGELOG.md)
|


## Features

#### JSON-RPC 2.0

* According to [JSON-RPC 2.0 Specifiation](http://www.jsonrpc.org/specification)
* However: [Batches](http://www.jsonrpc.org/specification#batch) are not yet
  supported.
* For the transport level the following message framing options are available:
    * `:none` - This is the default
       [frameless](http://www.simple-is-better.org/json-rpc/transport_sockets.html#pipelined-requests-responses-json-splitter) stream serialization mode.
    * `:rfc-7464` - Message framing according to
      [rfc-7464](https://tools.ietf.org/html/rfc7464) specification.

#### BSON-RPC

* Identical to [JSON-RPC 2.0 Specifiation](http://www.jsonrpc.org/specification)
  with following differences:
    * Messages are encoded as [BSON](http://bsonspec.org/spec.html) instead of
      JSON.
    * Protocol identifier is "bsonrpc" instead of "jsonrpc".
    * Batches are not supported since BSON does not have top-level arrays.
* Benefits over JSON-RPC:
    * Has a Binary type, without message size penalties.
        * JSON-RPC does not have binary type -> Must be carried as strings
          and requires schemas/metadata to custom-handle these fields.
        * Base64 encoded binary within JSON-RPC suffers 33% increase in size.
    * Has a Datetime type.

#### Transport

* This library uses [Manifold](https://github.com/ztellman/manifold) message
  streams. Documentation here implicitly assumes that the
  [aleph](https://github.com/ztellman/aleph) library will be used as the
  TCP (+TLS) connectivity provider, but any stream connection which can be
  wrapped into manifold duplex-stream should suffice.

#### Other

* Logs with [timbre](https://github.com/ptaoussanis/timbre).


## Quickstart

### Minimalistic example
#### TCP Server
```clojure
(ns my.ns
  (:require [aleph.tcp :as tcp]
            [clj-bson-rpc.core :as rpc]))

(def request-handlers
  {:swap-it (fn [msg] (apply str (reverse msg)))
   :intersperse (fn [c msg] (apply str (interpose c (seq msg))))})

(defn connection-handler [s info]
  ;; Start serving requests (in core.async go-block):
  (rpc/connect-json-rpc! s request-handlers {}))

(tcp/start-server connection-handler {:port 4321})
```
On the TCP-server side the client connections are promoted to RPC context with
`connect-json-rpc!` taking socket, request-handlers and an empty Map of notification
handlers. This call forks a core.async process which dispatches all incoming
requests, notifications and responses to their respectful handlers.
`connect-json-rpc!` returns a "context" object which is ignored in this example
server.

#### TCP Client
```clojure
(ns my.ns
  (:require [aleph.tcp :as tcp]
            [clj-bson-rpc.core :as rpc]))

(def rpc-ctx (-> {:host "localhost" :port 4321}
                 @(tcp/client)
                 (rpc/connect-json-rpc!)))

(println (rpc/request! rpc-ctx :swap-it "example"))
; elpmaxe

(println (rpc/request! rpc-ctx :intersperse "--" "example"))
; e--x--a--m--p--l--e
```
On the TCP-client side the connection is also wrapped with
`connect-json-rpc!` but this time we choose to not set any request nor
notification -handlers. The returned rpc-ctx is necessary for making
requests to the Peer Node (= server).

### Full example

#### TCP Server
```clojure
(ns my.ns
  (:require [aleph.tcp :as tcp]
            [clj-bson-rpc.core :as rpc]))

;; Some functions inteded to be exposed for clients to use.

(defn greetings
  [client content]
  (str "Hello" client ":" content))

(defn process
  [rpc-ctx a b c]
  (do
    (println a)
    ;; Send notification back to client during the processing
    (rpc/notify! rpc-ctx :report a)
    (println b)
    ;; Special function for closing connection to peer from
    ;; within request handler:
    (if (= a b) (rpc/close-connection! "Bye!"))
    (rpc/notify! rpc-ctx :report b)
    (println c)
    (rpc/notify! rpc-ctx :report c)
    "Ready"))

;; Set up RPC

(defn requests [client rpc-ctx]
  "client  - info about tcp peer
   rpc-ctx - RPC context to use for callbacks"
  {; When called from client the :greetings takes one argument.
   :greetings (partial greetings client)
   ; When called from client the :process takes 3 arguments.
   :process (partial process rpc-ctx)})

(defn connection-handler [s info]
  (let [requests (partial requests info)
        options {:json-framing :rfc-7464}]
    ; connect-json-rpc! detects that `requests` is a function and
    ; calls it internally with `rpc-ctx` in order to acquire
    ; fully defined interface. (see API doc)
    (rpc/connect-json-rpc! s requests {} options)))

(tcp/start-server connection-handler {:port 4321})
```

#### TCP Client
```clojure
(ns my.ns
  (:require [aleph.tcp :as tcp]
            [clj-bson-rpc.core :as rpc]))

;; Some functions intended to be exposed for server to use.

(defn demo-callback [rpc-ctx x] (+ x 32))

(defn report [item]
  (println "Reporting:" item))

;; Set up RPC

(defn requests [rpc-ctx]
  {:demo-callback (partial demo-callback rpc-ctx)})

(def notifications {:report report})

(def rpc-ctx (-> {:host "localhost" :port 4321}
                 @(tcp/client)
                 (rpc/connect-json-rpc! requests notifications {:json-framing :rfc-7464})))

(println "Finally" (rpc/request! rpc-ctx :process 11 22 33))
; Reporting: 11
; Reporting: 22
; Reporting: 33
; Finally Ready
```

### API - BSON or JSON

* Selection between BSON and JSON boils down to choosing either
  `connect-bson-rpc!` or `connect-json-rpc!`. Option differences
  are minimal and documented in the
  [API doc](http://seprich.github.io/clj-bson-rpc/codox/clj-bson-rpc.core.html).
* Other API functions are used identically for both BSON and JSON.


### TODO

* JSON-RPC Batch support
* Support dynamic modification of rpc method and notification handlers.
  Enabling e.g. register/unregister-method functionality.

## License

Copyright © 2015 Jussi Seppälä

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [unreleased]

## [0.2.5] - 2016-03-06
### Changed
- Update dependencies

## [0.2.4] - 2016-01-28
### Changed
- Update dependencies

## [0.2.3] - 2016-01-01
### Changed
- Dependency update

## [0.2.2] - 2016-01-01
### Changed
- Dependencies update

## [0.2.1] - 2015-12-09
### Fixed
- Upon invalid JSON either log only or close connection depending on
  the used `:json-framing` transport option.

## [0.2.0] - 2015-12-08
### Added
- JSON-RPC support: `connect-json-rpc!`

### Changed
- namespace `tcp` -> `core`
- Function `connect-rpc!` -> `connect-bson-rpc!`

## [0.1.0] - 2015-11-26
### Initial Version:
- BSON-RPC - requests and notifications over TCP connection.

[unreleased]: https://github.com/seprich/clj-bson-rpc/compare/0.2.5...HEAD
[0.2.5]: https://github.com/seprich/clj-bson-rpc/compare/0.2.4...0.2.5
[0.2.4]: https://github.com/seprich/clj-bson-rpc/compare/0.2.3...0.2.4
[0.2.3]: https://github.com/seprich/clj-bson-rpc/compare/0.2.2...0.2.3
[0.2.2]: https://github.com/seprich/clj-bson-rpc/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/seprich/clj-bson-rpc/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/seprich/clj-bson-rpc/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/seprich/clj-bson-rpc/tree/0.1.0


test/clj_bson_rpc/rpc_test.clj
(ns clj-bson-rpc.rpc-test
  (:require
    [clojure.test :refer :all]
    [clj-bson-rpc.rpc :refer :all]))

(def test-messages [;; Valid requests
                    {:id "req-1"
                     :bsonrpc "2.0"
                     :method "test-echo"
                     :params [1 2 3]}
                    {:id 22
                     :bsonrpc "2.0"
                     :method "test-echo"
                     :extra {:bar "bara"}}
                    ;; Invalid requests
                    {:id "req-3"
                     :bsonrpc "1.0"
                     :method "test-echo"
                     :params ["foo"]}
                    {:bsonrpc "2.0"
                     :params [1 2 3]}
                    ;; Valid response
                    {:id "resp-1"
                     :bsonrpc "2.0"
                     :result "hello"}
                    ;; Invalid response
                    {:id "resp-2"
                     :bsonrpc "2.0"
                     :result "hello"
                     :error "yello"}
                    ;; Valid error response
                    {:id "eresp-1"
                     :bsonrpc "2.0"
                     :error {:code -32000
                             :message "Server error"
                             :data "Backtrace foo"}}
                    ;; Valid nil id error response
                    {:id nil
                     :bsonrpc "2.0"
                     :error {:code -32700
                             :message "Parse error"}}
                    ;; Valid notifications
                    {:bsonrpc "2.0"
                     :method "notify"
                     :params ["foo"]}
                    {:bsonrpc "2.0"
                     :method "event-X"}
                    ;; Invalid error response
                    {:id "eresp-2"
                     :bsonrpc "2.0"
                     :error {:message "Not enough"}}
                    ;; Garbage messages
                    {}
                    {:id "some"}
                    {:id "other"
                     :jsonrpc "2.0"
                     :result 34}
                    ;;
                    ])

(def ctx  {:protocol-keyword :bsonrpc})

(defn match
  [t-fn expected]
  (doseq [[i msg] (map-indexed vector test-messages)]
    (if (t-fn msg)
      (is (some #{i} expected))
      (is (not (some #{i} expected))))))

(deftest recognize-request
  (let [request? (partial request? ctx)
        expected [0 1]]
    (match request? expected)))

(deftest recognize-notification
  (let [notification? (partial notification? ctx)
        expected [8 9]]
    (match notification? expected)))

(deftest recognize-response
  (let [response? (partial response? ctx)
        expected [4 6]]
    (match response? expected)))

(deftest recognize-nil-id-error-response
  (let [nil-id-error-response? (partial nil-id-error-response? ctx)
        expected [7]]
    (match nil-id-error-response? expected)))

(def notif? (partial notification? ctx))

(deftest notifications-without-handlers
  (let [notifications (filterv notif? test-messages)
        errors (atom [])
        err-handler (fn [e] (swap! errors (fn [v] (conj v e))))]
    (doseq [notification notifications]
      (handle-notification err-handler {} notification))
    (is (= (mapv ex-data @errors) [{:type :notification-handler-not-found :method "notify"}
                                   {:type :notification-handler-not-found :method "event-X"}]))))

(deftest notifications-with-handlers
  (let [notifications (filterv notif? test-messages)
        errors (atom [])
        err-handler (fn [e] (swap! errors (fn [v] (conj v e))))
        calls (atom [])
        handlers {"notify" (fn [par] (swap! calls (fn [v] (conj v (str "(notify " par ")")))))}]
    (doseq [notification notifications]
      (handle-notification err-handler handlers notification))
    (is (= (mapv ex-data @errors) [{:type :notification-handler-not-found :method "event-X"}]))
    (is (= @calls ["(notify foo)"]))))

(def req? (partial request? ctx))

(defn test-echo [a b c] (str "Hello " a ", " b ", " c))

(deftest handle-requests
  (let [requests (filterv req? test-messages)
        handlers {"test-echo" test-echo}
        responses (mapv #(handle-request :bsonrpc handlers %) requests)
        expected [{:bsonrpc "2.0"
                   :id "req-1"
                   :result "Hello 1, 2, 3"}
                  {:bsonrpc "2.0"
                   :id 22
                   :error {:code -32602
                           :message "Invalid params"
                           :data "clojure.lang.ArityException: Wrong number of args (0) passed to: rpc-test/test-echo"}}]]
    (is (= responses expected))))


test/clj_bson_rpc/tcp_test.clj
(ns clj-bson-rpc.tcp-test
  (:require
    [clojure.test :refer :all]
    [clj-bson-rpc.tcp :refer :all]
    [manifold.stream :as stream]
    [taoensso.timbre :as log]))

;(log/set-level! :trace)

(defn- create-duplex-stream []
  (let [a (stream/stream)
        b (stream/stream)]
    [(stream/splice a b) (stream/splice b a)]))

(def simple-request-handlers
  {:echo (fn [msg] (apply str (reverse msg)))
   :ping (fn [msg] (if (= msg "ping!")
                     "pong!"
                     (throw (ex-info "Value \"ping!\" expected!" {:type :test-error}))))
   :exit (fn [] (close-connection! "ack!"))})

(defn- gen-server-client! []
  (let [[a b] (create-duplex-stream)]
    [(connect-rpc! a simple-request-handlers {}) (connect-rpc! b)]))

(deftest simple-request
  (let [[s c] (gen-server-client!)]
    (is (= (request! c :echo "Hello!") "!olleH"))))

(deftest multiple-requests
  (let [[s c] (gen-server-client!)]
    (is (= (request! c :echo "Hi there!") "!ereht iH"))
    (is (= (request! c :ping "ping!") "pong!"))
    (is (= (request! c :echo "other") "rehto"))))

(deftest request-errors
  (let [[s c] (gen-server-client!)]
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Method not found"
                          (request! c :tak-ada "blam!")))
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Invalid params"
                          (request! c :echo "too" "many" "arguments")))
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Server error"
                          (request! c :ping "implemented error")))))

(deftest client-closes
  (let [[s c] (gen-server-client!)]
    (is (= (request! c :echo "Client closes!") "!sesolc tneilC"))
    (close! c)
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Connection closed"
                          (request! c :echo "nono")))))

(deftest server-closes
  (let [[s c] (gen-server-client!)]
    (is (= (request! c :echo "Server closes!") "!sesolc revreS"))
    (is (= (request! c :exit) "ack!"))
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Connection closed"
                          (request! c :echo "nono")))))


(defn- gen-request-handlers [rpc-ctx]
  {:process (fn [msg]
              (doseq [x msg] (notify! rpc-ctx :note x))
              "Done!")
   :echo (fn [msg] (apply str (reverse msg)))})

(defn- gen-notification-handlers [test-verify rpc-ctx]
  {:note (fn [msg] (swap! test-verify (fn [v] (conj v msg))))})

(deftest requests-and-notifications
  (let [record (atom [])
        [a b] (create-duplex-stream)
        ;; services in server
        s (connect-rpc! a gen-request-handlers {})
        ;; client which accepts notifications from server
        c (connect-rpc! b {} (partial gen-notification-handlers record))]
    (is (= (request! c :process "Whammy!") "Done!"))
    (is (= @record ["W" "h" "a" "m" "m" "y" "!"]))))


test/clj_bson_rpc/bson_test.clj
(ns clj-bson-rpc.bson-test
  (:require
    [clojure.test :refer :all]
    [clj-bson-rpc.bson :refer :all]
    [clj-bson-rpc.bytes :refer :all]
    [manifold.stream :as stream]))

(defn- bytes->hex
  [b]
  (apply str (mapv #(format "%02x" %) b)))

(defn- hex->bytes
  [h]
  (byte-array
    (mapv
      (fn [[x y]] (unchecked-byte (Integer/parseInt (str x y) 16)))
      (partition 2 h))))

(defn- slurp-bytes
  [filename]
  (with-open [out (java.io.ByteArrayOutputStream.)]
    (-> (clojure.java.io/input-stream filename)
        (clojure.java.io/copy out))
    (.toByteArray out)))

(defn- create-duplex-stream
  []
  (let [a (stream/stream)
        b (stream/stream)]
    [(stream/splice a b) (stream/splice b a)]))

(deftest simple-encoding
  (let [data {:a "foo" :b "bar"}
        expected (str "1b000000"  ; 27 bytes
                      "02""6100""04000000""666f6f00" ; type-str key=a len=4 foo\0
                      "02""6200""04000000""62617200" ; type-str key=b len=4 bar\0
                      "00")] ; 0
    (is (= (bytes->hex (encode data)) expected))))

(deftest simple-decoding
  (let [data (str "29000000" ; 41 bytes
                  "12""6100""0100000000000000" ; type-int64 key=a 1
                  "02""6200""07000000""7365636f6e6400"
                  "12""6300""0300000000000000"
                  "00")
        expected {:a 1 :b "second" :c 3}]
    (is (= (decode (hex->bytes data)) expected))))

(deftest roundtrip
  (let [data {:some 1 :items ["first" "second"] :nested {:foo "bar"}}]
    (is (= (decode (encode data)) data))))

(deftest encoding-tiny-binary
  (let [data {:test "some"
              :int 988
              :content (hex->bytes "000503abba")}
        expected (str "34000000"
                      "02""7465737400""05000000""736f6d6500"
                      "12""696e7400""dc03000000000000"
                      "05""636f6e74656e7400""05000000""00""000503abba"
                      "00")]
    (is (= (bytes->hex (encode data)) expected))))

(deftest binary-content
  (let [binary-data (slurp-bytes "test/data/penguin.png")
        data {:filename "penguin.png"
              :filesize 27466
              :content binary-data}
        encoded (encode data)
        head73 (str "896b0000"
                    "02""66696c656e616d6500""0c000000""70656e6775696e2e706e6700"
                    "12""66696c6573697a6500""4a6b000000000000"
                    "05""636f6e74656e7400""4a6b0000""00"
                        "89504e470d0a1a0a000000") ; etc
        decoded (decode encoded)]
    (is (= (bytes->hex (take 73 encoded)) head73))
    (is (= (:filename decoded) (:filename data)))
    (is (= (:filesize decoded) (:filesize data)))
    ; (= data decoded) does not work as expected because
    ;  byte array equality only compares objects themselves and not the
    ;  equality of the content bytes. Comparing the hexified contents however
    ;  proves content equality.
    (is (= (bytes->hex (:content decoded)) (bytes->hex (:content data))))))

(deftest bson-codec-decoding-simple
  (let [[a b] (create-duplex-stream)
        original {:a "foo" :b "bar" :c "barbara"}
        bson-stream (bson-codec b)]
    (is @(stream/put! a (encode original)))
    (is (= @(stream/take! bson-stream) original))))

(deftest bson-codec-decoding-dirty-close
  (let [[a b] (create-duplex-stream)
        original {:a "foo" :b "bar" :c "barbara"}
        full-raw (encode original)
        double-raw (bbuf-concat full-raw full-raw)
        [broken tail] (bbuf-split-at (- (count double-raw) 10) double-raw)
        bson-stream (bson-codec b)]
    (is @(stream/put! a broken))
    (stream/close! a)
    (is (= @(stream/take! bson-stream) original))
    (let [error @(stream/take! bson-stream)]
      (is (instance? clojure.lang.ExceptionInfo error))
      (is (= (:type (ex-data error)) :trailing-garbage)))
    (is (nil? @(stream/take! bson-stream)))))

(deftest bson-codec-decoding-garbage
  (let [[a b] (create-duplex-stream)
        garbage (byte-array [0x0c 0x00 0x00 0x00   ; len 12 bytes
                             0xb5 0xae 0xbc 0x67
                             0x22 0xa2 0x06 0x00])
        ok-bytes (encode {:a "bom" :b "dia"})
        bson-stream (bson-codec b)]
    (is @(stream/put! a (bbuf-concat garbage ok-bytes)))
    ;(stream/close! a)
    (let [error @(stream/take! bson-stream)]
      (is (instance? clojure.lang.ExceptionInfo error))
      (is (= (:type (ex-data error)) :invalid-bson)))
    (stream/close! a)
    (is (= @(stream/take! bson-stream) {:b "dia" :a "bom"}))
    (is (nil? @(stream/take! bson-stream)))))

(deftest bson-codec-decoding-framing-error
  (let [[a b] (create-duplex-stream)
        garbage (byte-array [0xdf 0xff 0xff 0xff   ; len -3 bytes
                             0x02 0x02 0x02 0x02])
        bson-stream (bson-codec b)]
    (is @(stream/put! a garbage))
    (let [error @(stream/take! bson-stream)]
      (is (instance? clojure.lang.ExceptionInfo error))
      (is (= (:type (ex-data error))) :invalid-framing)
      )
    ))

(deftest bson-codec-encoding-simple
  (let [[a b] (create-duplex-stream)
        original {:a "foo" :b "bar" :c "barbara"}
        bson-stream (bson-codec b)]
    (is @(stream/put! bson-stream original))
    (is (= (bytes->hex @(stream/take! a)) (bytes->hex (encode original))))))


project.clj
(defproject clj-bson-rpc "0.2.5"
  :description "bson rpc protocol"
  :url "http://github.com/seprich/clj-bson-rpc"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[com.taoensso/timbre "4.3.1" :exclusions [org.clojure/tools.reader]]
                 [congomongo "0.4.8"]
                 [manifold "0.1.2"]
                 [org.clojure/clojure "1.8.0"]
                 [org.clojure/core.async "0.2.374"]
                 [org.clojure/data.json "0.2.6"]] ; NOTE: not using cheshire due to issue #94.
  :plugins [[lein-ancient "0.6.8"] ; https://github.com/xsc/lein-ancient/issues/58
            [lein-codox "0.9.4"]
            [lein-kibit "0.1.2"]
            [michaelblume/lein-marginalia "0.9.0"]
            [lein-ns-dep-graph "0.1.0-SNAPSHOT"]]
  :codox {:namespaces [clj-bson-rpc.core]
          :source-uri "https://github.com/seprich/clj-bson-rpc/blob/master/{filepath}#L{line}"
          :metadata {:doc/format :markdown}}
  :profiles {:debug {:debug true
                     :injections [(prn (into {} (System/getProperties)))]}
             :dev {}})


src/clj_bson_rpc/bson.clj
(ns clj-bson-rpc.bson
  "BSON encoding and decoding. Utilizes mongodb codecs for
   clojure <-> bson conversions.
  "
  (:import
    [org.bson BasicBSONDecoder BasicBSONEncoder])
  (:require
    [clj-bson-rpc.bytes :refer [bbuf-concat bbuf-split-at peek-int32-le]]
    [manifold.stream :as stream]
    [somnium.congomongo.coerce :refer [coerce]]))

(defn encode
  "Encode message to BSON byte-array"
  [m]
  (.encode (BasicBSONEncoder.) (coerce m [:clojure :mongo])))

(defn decode
  "Decode BSON byte-array to message"
  [b]
  (coerce (.readObject (BasicBSONDecoder.) b) [:mongo :clojure]))

(defn- stream-decoder
  "Lift BSON message(s) from source stream `s` based on BSON frame recognition,
   decode to clojure (Map messages) and push to the result stream. Various errors will
   produce `ex-info` items to the result stream.
  "
  [s max-len]
  (let [source (stream/->source s)
        sink (stream/stream)
        find-frames (fn [bbuff]
                      (loop [frames [] remainder bbuff]
                        (let [blen (count remainder)
                              rlen (when (>= blen 4) (peek-int32-le remainder))
                              err-long #(ex-info (str "Message length exceeds the maximum allowed. "
                                                      "Max: " max-len ", Message len: " rlen)
                                                 {:type :exceeds-max-length :bytes remainder})
                              err-frame #(ex-info "Message framing error."
                                                  {:type :invalid-framing :bytes remainder})]
                          (cond
                            (nil? rlen)      [frames remainder]
                            (> rlen max-len) [(conj frames (err-long)) remainder]
                            (< rlen 5)       [(conj frames (err-frame)) remainder]
                            (>= blen rlen)   (let [[frame remainder] (bbuf-split-at rlen remainder)]
                                               (recur (conj frames frame) remainder))
                            :else            [frames remainder]))))
        decode-frame (fn [frame]
                       (if (instance? clojure.lang.ExceptionInfo frame)
                         frame
                         (try
                           (decode frame)
                           (catch Exception e
                             (ex-info (str e) {:type :invalid-bson :bytes frame})))))
        buffer (atom (byte-array []))
        decoder (fn [new-bytes]
                  (let [[frames remainder] (find-frames (bbuf-concat @buffer new-bytes))
                        trail (if (and (stream/drained? source) (seq remainder))
                                [(ex-info "Non-decodable trailing bytes when stream was closed."
                                          {:type :trailing-garbage :bytes remainder})]
                                [])]
                    (reset! buffer remainder)
                    (->> (into frames trail)
                         (mapv decode-frame)
                         (stream/put-all! sink))))]
    (stream/connect-via source decoder sink {:downstream? false})
    (stream/on-drained source #(do (decoder []) (stream/close! sink)))
    sink))

(defn bson-codec
  "BSON codec for a manifold duplex stream.

   * `s` - Duplex stream of raw bytes (possibly from aleph)
   * `max-len` - Optional argument for limiting incoming message maximum size.

   Returns a duplex stream which:

    * As a sink takes clojure maps which are encoded and sent to the byte stream `s`
    * As a source produces decoded (clojurized) BSON messages from `s`

   Any non-decodable byte sequences from the raw stream are converted to
   clojure.lang.ExceptionInfo objects which are takeable from the returned
   duplex stream. Thus take! from this source will either:

    * Block waiting a new message.
    * Return a bson message.
    * Return a clojure.lang.ExceptionInfo object.
      * Message framing error (bson length is negative or < 5)-> :type :invalid-framing
      * Message decoding error from bson parser-> :type :invalid-bson
        ! Stream `s` must be closed.
      * Non-decodable trailing bytes when stream closed-> :type :trailing-garbage
    * Return nil if underlying stream `s` is closed and source is drained.
  "
  [s & {:keys [max-len] :or {max-len (Integer/MAX_VALUE)}}]
  (let [out (stream/stream)]
    (stream/connect (stream/map encode out) s)
    (stream/splice out (stream-decoder s max-len))))


src/clj_bson_rpc/rpc.clj
(ns clj-bson-rpc.rpc
  "BSON-RPC Connection over [manifold](https://github.com/ztellman/manifold) duplex stream.

   Practical applications include (but limited to) using [aleph](https://github.com/ztellman/aleph)
   to establish TCP connection (+ TLS) between the client and the server.

   After an RPC connection has been set on both sides, the RPC nodes are principally equal.
   The TCP client may provide services which the TCP server then uses and vice versa.
  "
  )

;; ## Constants and Helpers

(def rpc-error
  {:parse-error      {:code -32700 :message "Parse error"}
   :invalid-request  {:code -32600 :message "Invalid Request"}
   :method-not-found {:code -32601 :message "Method not found"}
   :invalid-params   {:code -32602 :message "Invalid params"}
   :internal-error   {:code -32603 :message "Internal error"}
   :server-error     {:code -32000 :message "Server error"}})

(def rpc-version "2.0")

(defonce identifier-counter (atom 1))
(defn default-id-generator [] (str "id-" (swap! identifier-counter inc)))

;; ## Message Type Identifiers

(defn request?
  [{:keys [protocol-keyword]} msg]
  (and (= (get msg protocol-keyword) rpc-version)
       (string? (:method msg))
       (contains? msg :id) (or (string? (:id msg)) (integer? (:id msg)) (nil? (:id msg)))))

(defn notification?
  [{:keys [protocol-keyword]} msg]
  (and (= (get msg protocol-keyword) rpc-version)
       (string? (:method msg))
       (not (contains? msg :id))))

(defn response?
  [{:keys [protocol-keyword]} msg]
  (let [result-and-no-error (and (contains? msg :result) (not (contains? msg :error)))
        error-and-no-result (and (not (contains? msg :result))
                                 (integer? (get-in msg [:error :code]))
                                 (string? (get-in msg [:error :message])))]
    (and (= (get msg protocol-keyword) rpc-version)
         (contains? msg :id) (or (string? (:id msg)) (integer? (:id msg)))
         (or result-and-no-error error-and-no-result))))

(defn nil-id-error-response?
  [{:keys [protocol-keyword]} msg]
  (let [error-and-no-result (and (not (contains? msg :result))
                                 (integer? (get-in msg [:error :code]))
                                 (string? (get-in msg [:error :message])))]
    (and (= (get msg protocol-keyword) rpc-version)
         (contains? msg :id) (nil? (:id msg))
         error-and-no-result)))

;; ## Inbound Requests, Notifications

(defn handle-request
  [protocol-keyword request-handlers msg]
  (let [method-name-kw (keyword (:method msg))
        method-name (name (:method msg))
        params (:params msg)
        id (:id msg)
        ok-response (fn [result] {protocol-keyword rpc-version
                                  :result result
                                  :id id})
        invalid-params (fn [e-msg] {protocol-keyword rpc-version
                                    :error (assoc (:invalid-params rpc-error) :data e-msg)
                                    :id id})
        method-not-found (fn [] {protocol-keyword rpc-version
                                 :error (:method-not-found rpc-error)
                                 :id id})
        request-handlers (if (instance? clojure.lang.IDeref request-handlers)
                           (deref request-handlers)
                           request-handlers)
        send-response (fn [method params]
                        (try
                          (ok-response (apply (get request-handlers method) params))
                          (catch clojure.lang.ArityException e (invalid-params (str e)))))]
    (if (contains? request-handlers method-name-kw)
      (send-response method-name-kw params)
      (if (contains? request-handlers method-name)
        (send-response method-name params)
        (method-not-found)))))

(defn handle-notification
  [notification-error-handler notification-handlers msg]
  (let [method-name-kw (keyword (:method msg))
        method-name (name (:method msg))
        params (:params msg)
        notification-handlers (if (instance? clojure.lang.IDeref notification-handlers)
                                (deref notification-handlers)
                                notification-handlers)]
    (if (contains? notification-handlers method-name-kw)
      (apply (get notification-handlers method-name-kw) params)
      (if (contains? notification-handlers method-name)
        (apply (get notification-handlers method-name) params)
        (notification-error-handler
          (ex-info (str "No handler found for notification:" method-name)
                   {:type :notification-handler-not-found :method method-name}))))))
